core/src/main/java/org/elasticsearch/index/analysis/KeywordMarkerTokenFilterFactory.java

```diff
@@ -21,32 +21,69 @@
 
 import org.apache.lucene.analysis.CharArraySet;
 import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.miscellaneous.PatternKeywordMarkerFilter;
 import org.apache.lucene.analysis.miscellaneous.SetKeywordMarkerFilter;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.index.IndexSettings;
 
 import java.util.Set;
+import java.util.regex.Pattern;
 
+/**
+ * A factory for creating keyword marker token filters that prevent tokens from
+ * being modified by stemmers.  Two types of keyword marker filters are available:
+ * the {@link SetKeywordMarkerFilter} and the {@link PatternKeywordMarkerFilter}.
+ *
+ * The {@link SetKeywordMarkerFilter} uses a set of keywords to denote which tokens
+ * should be excluded from stemming.  This filter is created if the settings include
+ * `keywords`, which contains the list of keywords, or `keywords_path`, which
+ * contains a path to a file in the config directory with the keywords.
+ *
+ * The {@link PatternKeywordMarkerFilter} uses a regular expression pattern to match
+ * against tokens that should be excluded from stemming.  This filter is created if
+ * the settings include `keywords_pattern`, which contains the regular expression to
+ * match against.
+ */
 public class KeywordMarkerTokenFilterFactory extends AbstractTokenFilterFactory {
 
     private final CharArraySet keywordLookup;
+    private final Pattern keywordPattern;
 
     public KeywordMarkerTokenFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) {
         super(indexSettings, name, settings);
 
         boolean ignoreCase =
             settings.getAsBooleanLenientForPreEs6Indices(indexSettings.getIndexVersionCreated(), "ignore_case", false, deprecationLogger);
-        Set<?> rules = Analysis.getWordSet(env, indexSettings.getIndexVersionCreated(), settings, "keywords");
-        if (rules == null) {
-            throw new IllegalArgumentException("keyword filter requires either `keywords` or `keywords_path` to be configured");
+        String patternString = settings.get("keywords_pattern");
+        if (patternString != null) {
+            // a pattern for matching keywords is specified, as opposed to a
+            // set of keyword strings to match against
+            if (settings.get("keywords") != null || settings.get("keywords_path") != null) {
+                throw new IllegalArgumentException(
+                    "cannot specify both `keywords_pattern` and `keywords` or `keywords_path`");
+            }
+            keywordPattern = Pattern.compile(patternString);
```

Comment:
Good point.  However, there is no version of this filter that takes Lucene's regexes.  We would have to create a new filter in Lucene or change the `PatternKeywordMarkerFilter` to take Lucene regexes instead of the Java regexes. 

Another option is a special filter to just prevent stemming on words that are under a certain length (as requested in https://github.com/elastic/elasticsearch/issues/4877), but I imagine a regex filter here would be more generally useful as a solution, plus Lucene already has this.  The downside is that we could end up killing nodes processing strangely formatted documents with ill-conceived regexes.  Yet another option is to use the Lucene `PatternKeywordMarkerFilter` under the covers as is, but only expose the option to pass in a minimum keyword length in ES, and we convert that to a regular expression and use the `PatternKeywordMarkerFilter` to achieve the desired functionality.

@jpountz what do you think?